File: generated_code/GeminiPro/Java/025_SplayTree.java
/**
 * A self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-adjusting binary search tree with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private class Node {
        int key;
        Node left, right;

        public Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree remains unchanged.
     * After insertion, the new node becomes the root of the tree.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        // Case 1: The tree is empty.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the node with the key,
        // or the last-accessed node (the parent of where the key would be), to the root.
        root = splay(root, key);

        // If the key is already in the tree, we are done.
        if (root.key == key) {
            return;
        }

        // Case 2: The key is not in the tree. Insert the new node as the new root.
        Node newNode = new Node(key);
        if (key < root.key) {
            // The new root's key is smaller than the old root's key.
            // The new root's right child becomes the old root.
            // The new root's left child becomes the old root's left child.
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            // The new root's key is larger than the old root's key.
            // The new root's left child becomes the old root.
            // The new root's right child becomes the old root's right child.
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Searches for a key in the splay tree.
     * If the key is found, the node containing the key is splayed to the root.
     * If the key is not found, the last accessed node in the search path is splayed to the root.
     *
     * @param key The integer key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is not in the tree, the tree remains unchanged.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree for the key. If the key exists, it becomes the root.
        root = splay(root, key);

        // If the key is not at the root, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // The key is at the root. Now we need to remove it.
        // The tree is split into two subtrees: left and right.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to its root.
            // Since `key` is larger than any element in the left subtree, splaying for `key`
            // will bring the maximum element to the root of the left subtree.
            leftSubtree = splay(leftSubtree, key);

            // The original right subtree becomes the right child of the new root.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }

    /**
     * Performs a right rotation on the subtree rooted at y.
     *
     * @param y The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node rightRotate(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. It brings the node with the given key to the root of the
     * subtree rooted at `node`. If the key is not present, it brings the last accessed
     * node to the root.
     *
     * @param node The root of the current subtree.
     * @param key The key to splay for.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node node, int key) {
        // Base cases: node is null or the key is at the root.
        if (node == null || node.key == key) {
            return node;
        }

        // Key is in the left subtree.
        if (key < node.key) {
            // If the left child is null, the key is not in the tree.
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left-Left)
            if (key < node.left.key) {
                // Recursively bring the key to the root of the left-left grandchild.
                node.left.left = splay(node.left.left, key);
                // Perform the first right rotation on the current node (g).
                node = rightRotate(node);
            }
            // Zig-Zag (Left-Right)
            else if (key > node.left.key) {
                // Recursively bring the key to the root of the left-right grandchild.
                node.left.right = splay(node.left.right, key);
                // Perform the first rotation (left) on the child (p).
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }
            // Perform the second rotation (for Zig-Zig/Zig-Zag) or the only rotation (for Zig).
            return (node.left == null) ? node : rightRotate(node);

        }
        // Key is in the right subtree.
        else { // key > node.key
            // If the right child is null, the key is not in the tree.
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right-Left)
            if (key < node.right.key) {
                // Recursively bring the key to the root of the right-left grandchild.
                node.right.left = splay(node.right.left, key);
                // Perform the first rotation (right) on the child (p).
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right-Right)
            else if (key > node.right.key) {
                // Recursively bring the key to the root of the right-right grandchild.
                node.right.right = splay(node.right.right, key);
                // Perform the first left rotation on the current node (g).
                node = leftRotate(node);
            }
            // Perform the second rotation (for Zig-Zig/Zig-Zag) or the only rotation (for Zig).
            return (node.right == null) ? node : leftRotate(node);
        }
    }
}

File: generated_code/GeminiPro/Python/025_SplayTree.py
class SplayTree:
    """
    A complete, self-contained Splay Tree class that implements a
    dictionary-like set for integers.

    This data structure provides amortized O(log n) time complexity for search,
    insert, and delete operations. The key feature is the splaying operation,
    which moves frequently accessed elements closer to the root of the tree.
    """

    class _Node:
        """A private helper class for the nodes of the tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty splay tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splaying operation on node x, moving it to the root.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left:
                if parent == grandparent.left:  # Zig-Zig case
                    self._right_rotate(grandparent)
                    self._right_rotate(parent)
                else:  # Zig-Zag case
                    self._right_rotate(parent)
                    self._left_rotate(grandparent)
            else: # x is right child
                if parent == grandparent.right:  # Zig-Zig case
                    self._left_rotate(grandparent)
                    self._left_rotate(parent)
                else:  # Zig-Zag case
                    self._left_rotate(parent)
                    self._right_rotate(grandparent)

    def insert(self, key):
        """
        Inserts a key into the splay tree.
        If the key already exists, the node is splayed to the root.
        """
        node = self.root
        parent = None
        
        while node:
            parent = node
            if key == node.key:
                # Key already exists, splay it and we are done
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        new_node = self._Node(key)
        new_node.parent = parent
        
        if not parent:
            self.root = new_node  # Tree was empty
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
            
        # Splay the newly inserted node to the root
        self._splay(new_node)

    def search(self, key):
        """
        Searches for a key in the splay tree.
        
        Performs the splaying operation on the accessed node (if found) or
        its parent (if not found) to move it to the root.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        current = self.root
        last_visited = None
        
        while current:
            last_visited = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # If key is not found, splay the last visited node
        if last_visited:
            self._splay(last_visited)
            
        return False

    def delete(self, key):
        """
        Deletes a key from the splay tree.
        If the key is not found, the tree is splayed based on the search path.
        """
        # Splay the node with the key (or its would-be parent) to the root
        if not self.search(key):
            # Key was not found. search() already splayed the last accessed node.
            return
        
        # At this point, the node to delete is the root, and we know it exists
        # because search() would have returned False otherwise.
        if self.root.key != key:
            return

        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # No left child, promote the right subtree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        elif not right_subtree:
            # No right child, promote the left subtree
            self.root = left_subtree
            if left_subtree:
                left_subtree.parent = None
        else:
            # Both children exist. Join the two subtrees.
            left_subtree.parent = None # Disconnect for now
            
            # Find the maximum element in the left subtree.
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this maximum element to be the root of the left subtree.
            # We can do this by temporarily setting self.root to the left
            # subtree and then calling our existing _splay method.
            self.root = left_subtree
            self._splay(max_node)
            
            # After splaying, max_node is the new root (self.root).
            # It's guaranteed to not have a right child.
            # Attach the original right subtree.
            self.root.right = right_subtree
            right_subtree.parent = self.root